services/command-router/src/main/java/org/eclipse/hono/commandrouter/impl/CommandRouterServiceImpl.java

```diff
@@ -0,0 +1,394 @@
+        setSpecificConfig(configuration);
+    }
+
+    /**
+     * Sets the OpenTracing {@code Tracer} to use for tracking the processing
+     * of messages published by devices across Hono's components.
+     * <p>
+     * If not set explicitly, the {@code NoopTracer} from OpenTracing will
+     * be used.
+     *
+     * @param opentracingTracer The tracer.
+     */
+    @Autowired(required = false)
+    public final void setTracer(final Tracer opentracingTracer) {
+        LOG.info("using OpenTracing Tracer implementation [{}]", opentracingTracer.getClass().getName());
+        this.tracer = Objects.requireNonNull(opentracingTracer);
+    }
+
+    /**
+     * Sets the client for accessing device connection data.
+     *
+     * @param deviceConnectionInfo The client object.
+     * @throws NullPointerException if deviceConnectionInfo is {@code null}.
+     */
+    @Autowired
+    public final void setDeviceConnectionInfo(final DeviceConnectionInfo deviceConnectionInfo) {
+        this.deviceConnectionInfo = Objects.requireNonNull(deviceConnectionInfo);
+    }
+
+    /**
+     * Sets the client to use for accessing the Device Registration service.
+     *
+     * @param client The client.
+     * @throws NullPointerException if the client is {@code null}.
+     */
+    @Qualifier(RegistrationConstants.REGISTRATION_ENDPOINT)
+    @Autowired
+    public final void setRegistrationClient(final DeviceRegistrationClient client) {
+        this.registrationClient = Objects.requireNonNull(client);
+    }
+
+    /**
+     * Sets the factory to use for creating clients to receive commands.
+     *
+     * @param factory The factory.
+     * @throws NullPointerException if factory is {@code null}.
+     */
+    @Autowired
+    public final void setCommandConsumerFactory(final CommandConsumerFactory factory) {
+        this.commandConsumerFactory = Objects.requireNonNull(factory);
+    }
+
+    /**
+     * Sets the component for mapping an incoming command to the gateway (if applicable)
+     * and protocol adapter instance that can handle it.
+     *
+     * @param commandTargetMapper The mapper component.
+     * @throws NullPointerException if commandTargetMapper is {@code null}.
+     */
+    @Autowired
+    public final void setCommandTargetMapper(final CommandTargetMapper commandTargetMapper) {
+        this.commandTargetMapper = Objects.requireNonNull(commandTargetMapper);
+    }
+
+    @Override
+    public void start(final Promise<Void> startPromise) throws Exception {
+
+        if (registrationClient == null) {
+            startPromise.fail(new IllegalStateException("Device Registration client must be set"));
+        } else if (deviceConnectionInfo == null) {
+            startPromise.fail(new IllegalStateException("Device Connection info client must be set"));
+        } else {
+            startServiceClient(registrationClient, "Device Registration service");
+            if (deviceConnectionInfo instanceof ConnectionLifecycle) {
```

Comment:
*CacheBasedDeviceConnectionInfo* is no longer an instance of *ConnectionLifecycle* but is an instance of *org.eclipse.hono.util.Lifecylce* instead ...